pyomo/contrib/pynumero/interfaces/ampl_nlp.py

```diff
@@ -440,14 +440,24 @@ def get_constraints_scaling(self):
     def get_eq_constraints_scaling(self):
         constraints_scaling = self.get_constraints_scaling()
         if constraints_scaling is not None:
-            return np.compress(self._con_full_eq_mask, constraints_scaling)
+            # It is possible that derived classes may implement
+            # additional constraints.  We will assume that the first
+            # _n_con_full entries from get_constraints_scaling
+            # correspond to the constraints from AslNLP.
+            return np.compress(self._con_full_eq_mask,
+                               constraints_scaling[:self._n_con_full])
```

Comment:
This seems strange to me. If this method is problematic for derived classes, shouldn't they just override this method?

Reply:
Yes.  The original implementation of `PyomoGreyBoxNLP` inherited from `PyomoNLP` and just "tagged on" the grey box constraints / variables to the end of the base pyomo components.  The derived class did indeed overrode this method.  This implementation (and the associated comment) was done this way so that derived classes did not have to directly access private attributes from the base class.  That is, the `PyomoGreyBoxNLP.get_eq_constraints_scaling()` looked something like:
```python
def get_eq_constraints_scaling(self):
    return np.concatenate(
        super(PyomoGreyBoxNLP, self).get_eq_constriaints_scaling(),
        self._greybox_eq_con_scaling
    )
```

However, during development, we ran across other issues where some methods did not behave cleanly when subclassed, to the implementation of `PyomoGreyBoxNLP` in this PR uses containment instead of inheritance.  I would like to revisit the implementation / data structures in `PyomoNLP` so that it can more cleanly be subclassed, but that will wait for a future PR.